staging/src/k8s.io/endpointslice/topologyheuristics/manager.go

```diff
@@ -0,0 +1,344 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topologyheuristics
+
+import (
+	"fmt"
+	"slices"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/klog/v2"
+)
+
+// Manager manages multiple EndpointSlice Topology Heuristics. It exposes
+// functions for adding and removing topology hints to endpoint slices using
+// various heuristics.
+type Manager struct {
+	// heuristicsByName maintains a mapping of a heuristic name to a Heuristic
+	// instance.
+	heuristicsByName map[string]Heuristic
+	// defaultHeuristicName is the name of the Heuristic which will be considered
+	// the default.
+	defaultHeuristicName string
+	// nonEPSHintHeuristicNames should contain the names of heuristics which are not
+	// implemented through EndpointSlice hints and may be implemented in the
+	// Dataplane (or in other ways). Manager will NOT send "unrecognizable" events
+	// for these heuristic names (despite the fact that it does not populate
+	// EndpointSlice hints for these.)
+	nonEPSHintHeuristicNames []string
```

Comment:
This is suspicious to me - do we think we can enumerate all heuristics, or is it something that implementations can extend?

Reply:
That's a great question. I'll need some insights on helps on properly addressing this.

Here's what I am thinking:
1. It would be be unlikely to think that we can enumerate all heuristics here. 
2. @robscott, @aojea Based on the suggestion in 
![image](https://github.com/kubernetes/kubernetes/assets/26463869/92d614ff-7324-4484-9ec9-d817a98ce6a7) given that we now can't board 1.29 and thus aren't constrained by time, do we maybe want to add this to the KEP first? Perhaps something like the default EPS controller recognizes `Auto`, `Disabled` and `netorking.k8s.io/XYZHeuristic`?